### ISSUE TITLE
In some circumstances, canvas not being cleared properly

### ISSUE THREAD

**Issue by starwed** [craftyjs/Crafty#548]
As [reported by sulter](https://mail.google.com/mail/u/0/?shva=1#inbox/1410d711a25eace4) in the forum.

[This fiddle](http://jsfiddle.net/UmEn4/1/) reproduces the bug for me in both Firefox and Chrome.  

Probably expanding the cleared area by a pixel in each direction would fix this, but I'd like to understand _why_ it's happening.

**Comment by starwed** [comment]
After looking through the drawing code, I notice that the coordinates of dirty areas are converted to integers when regions overlap, but aren't in other cases.  This would explain why the behavior appears when the entity moves quickly -- when moving slowly, the old and new position would overlap, but when moving quickly they are disjoint.  

_e:_ Yes, coercing all coordinates to integers seems to work; just need to figure out the best/fastest way to do this.  A workaround in existing versions of crafty would be to make sure `x` and `y` are always integers.  (I was doing that as a matter of habit when using crafty, which is why I never noticed this issue!)

**Comment by starwed** [comment]
I think the solution will be to "round up" the dirty areas at the point that we render that region.  Currently there's some duplication of effort when areas get merged, and it's a bit messy to catch every case.  Far simpler to only round right before the point where it matters.

**Comment by Sulter** [comment]
Thanks for taking it up starwed. I realized myself that the issue was trying to set the x and y to float points. For some reason I just assumed that would be handled by the engine. 
I think it would be reasonable to just warn the users in the API, like it's done with the rotation. At least for now.

**Comment by starwed** [comment]
It certainly _should_ be handled by the engine!  :)  I'll try to land a fix for this tomorrow -- it's not very complicated, but I've had a vicious cold all week.

**Comment by starwed** [comment]
Well, I fixed the issue of the canvas not clearing.

I haven't landed it due to a lingering issue with the MBR, though -- unfortunately when rotating by exactly 90 degrees we get a little bit of slop due to trig functions not quite being zero.  (`Cos(Math.PI/2)` !== 0, for instance -- it's something like 10^-17 instead.)  So a square rotated by 90 degrees ends up increasing in size by 1 pixel.  This then causes a couple of tests to fail.

Obviously I can just alter the tests, but thought I'd bring it up here.  It's annoying that a rotation by 90/180/etc degrees changes things it shouldn't, but is this a serious problem that's worth fixing?  Or just something to live with?

**Comment by sbyrnes321** [comment]
It probably doesn't matter as far as I know. But isn't it an easy fix? The `_rotate` code calculates the sin and cosine...

``` javascript
    _rotate: function (v) {
        var theta = -1 * (v % 360), //angle always between 0 and 359
            rad = theta * DEG_TO_RAD,
            ct = Math.cos(rad), //cache the sin and cosine of theta
            st = Math.sin(rad),
            o = {
            x: this._origin.x + this._x,
            y: this._origin.y + this._y
        };

        //if the angle is 0 and is currently 0, skip
        if (!theta) {
            this._mbr = null;
            if (!this._rotation % 360) return;
        }
        // ------------- STEVE'S NEW PART ---------
        //if the angle is very very close to a multiple of 90 degrees, it is probably supposed to be exact,
        //so set cosine and sine to exactly 0 or 1 or -1.
        if (Math.abs(theta % 90) < 1e-10 || (90 - Math.abs(theta % 90)) < 1e-10) {
            ct = Math.round(ct);
            st = Math.round(st);
        }
```

Is that enough? Alternatively, could define an internal Crafty.cos_with_snap() and Crafty.sin_with_snap() that snap to 1,0,-1...

Sorry in advance if it's trickier than I realize. :-D

**Comment by starwed** [comment]
It really seems to just be an issue for "almost zero" numbers, so we could do something like

``` javascript
ct = (ct < 1e-10 && ct > -1e-10) ? 0 : ct
```

(and the same for st).

I don't really like doing this sort of special casing, but I guess it at least wouldn't have much speed impact.  (There are already lots of calculations going on when rotating, and this should be quite fast.)  Math.round and the like are still slow in IE 10, and so should probably be avoided for now.  (That's why we do weird bit fiddling things to find floor and ceiling when the code is going to be called a lot.)

**Comment by starwed** [comment]
This should now be fixed in develop by #550.
